### PR TITLE
guardrails model support

### PR DESCRIPTION
**Pull request by jngz-es**
### Description
Support remote guardrails model to validate user input or LLM output.
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/2463
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jngz-es** [comment]
Hi team, I am adding more UT and IT in this pr, you can review the feature at first.

File: common/src/main/java/org/opensearch/ml/common/model/Guardrail.java [hunk 1]
```diff
@@ -1,105 +1,18 @@
-/*
- * Copyright OpenSearch Contributors
- * SPDX-License-Identifier: Apache-2.0
- */
-
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/Guardrail.java:5]: Good catch, added.

File: common/src/main/java/org/opensearch/ml/common/model/Guardrail.java [hunk 2]
```diff
@@ -1,105 +1,18 @@
...
-    public static Guardrail parse(XContentParser parser) throws IOException {
-        List<StopWords> stopWords = null;
-        String[] regex = null;
+    public abstract void writeTo(StreamOutput out) throws IOException;
 
-        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
-        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
-            String fieldName = parser.currentName();
-            parser.nextToken();
+    public abstract Boolean validate(String input, Map<String, String> parameters);
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/Guardrail.java:15]: I prefer to use object not primitive, as long as memory saving is not critical.

File: common/src/main/java/org/opensearch/ml/common/model/Guardrails.java [hunk 3]
```diff
@@ -93,20 +111,47 @@ public static Guardrails parse(XContentParser parser) throws IOException {
...
                 .type(type)
-                .inputGuardrail(inputGuardrail)
-                .outputGuardrail(outputGuardrail)
+                .inputGuardrail(createGuardrail(type, inputGuardrailMap))
+                .outputGuardrail(createGuardrail(type, outputGuardrailMap))
                 .build();
     }
+
+    private static Boolean validateType(String type) {
+        Set<String> types = Set.of("local_regex", "model");
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/Guardrails.java:136]: Good point, changed.

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                        .build()
+        );
+        client
+                .execute(
+                        MLPredictionTaskAction.INSTANCE,
+                        request,
+                        new LatchedActionListener(actionListener, latch)
+                );
+        try {
+            latch.await(5, SECONDS);
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java:135]: Good question, maybe returning true is better than an exception.

[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java:135]: Removed the exception throwing.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        this.stopWords = stopWords;
+        this.regex = regex;
+    }
+    public LocalRegexGuardrail(@NonNull Map<String, Object> params) {
+        Object words = params.get(STOP_WORDS_FIELD);
+        stopWords = new ArrayList<>();
+        for (Map e : (List<Map>)words) {
+            stopWords.add(new StopWords(e));
+        }
+        this.regex = ((List<String>) params.get(REGEX_FIELD)).toArray(new String[0]);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:72]: Add null check for `params.get(REGEX_FIELD)` and `params.get(STOP_WORDS_FIELD)` ?

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+    public LocalRegexGuardrail(StreamInput input) throws IOException {
+        if (input.readBoolean()) {
+            stopWords = new ArrayList<>();
+            int size = input.readInt();
+            for (int i=0; i<size; i++) {
+                stopWords.add(new StopWords(input));
+            }
+        }
+        regex = input.readStringArray();
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:83]: The regex is optional, right ?

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            latch.countDown();
+            hitStopWords.set(true);
+        } finally {
+            if (context != null) {
+                context.close();
+            }
+        }
+
+        try {
+            latch.await(5, SECONDS);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:254]: Suggest change to non-blocking way with action listener. Which thread pool this part will block?

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            if (isStopWordsSystemIndex(indexName)) {
+                context = client.threadPool().getThreadContext().stashContext();
+                ThreadContext.StoredContext finalContext = context;
+                client.search(searchRequest, ActionListener.runBefore(new LatchedActionListener(ActionListener.<SearchResponse>wrap(r -> {
+                    if (r == null || r.getHits() == null || r.getHits().getTotalHits() == null || r.getHits().getTotalHits().value == 0) {
+                        hitStopWords.set(true);
+                    }
+                }, e -> {
+                    log.error("Failed to search stop words index {}", indexName, e);
+                    hitStopWords.set(true);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:231]: Why still set `hitStopWords` as true when exceptions happens ?

File: common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java:135]: Same comment here, best to use non-blocking way

File: common/src/main/java/org/opensearch/ml/common/model/StopWords.java [hunk 9]
```diff
@@ -33,6 +36,10 @@ public StopWords(String index, String[] sourceFields) {
         this.sourceFields = sourceFields;
     }
 
+    public StopWords(@NonNull Map<String, Object> params) {
+        this((String) params.get(INDEX_NAME_FIELD), ((List<String>) params.get(SOURCE_FIELDS_FIELD)).toArray(new String[0]));
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/StopWords.java:40]: Add null check `params.get(SOURCE_FIELDS_FIELD)`

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        }
+        return true;
+    }
+
+    public Boolean validateRegex(String input, Pattern pattern) {
+        Matcher matcher = pattern.matcher(input);
+        return !matcher.matches();
+    }
+
+    public Boolean validateStopWords(String input, Map<String, List<String>> stopWordsIndices) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:190]: Function name is not so easy to understand. Add some comment or change name ?

File: common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/ModelGuardrail.java:135]: Yeah, I will change it in next pr.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 5 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:72]: done.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 6 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:83]: changed to optional

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 7 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:254]: Will address it in next pr.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 5 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:72]: Added UT.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 6 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:83]: Added UT.

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 8 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:231]: Added comments.

File: common/src/main/java/org/opensearch/ml/common/model/StopWords.java (hunk 9 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/StopWords.java:40]: added

File: common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java (hunk 10 above)
Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java:190]: added

**Review by ylwu-amzn — APPROVED** [review]

**Review by b4sjoo — APPROVED** [review]

**Pull request merged**